Fix the handling of semver ranges that contain '*'

dashea opened the pull request:
Use the semver library to parse version strings into ranges. For any
version strings that fail to parse or are returned as '*', output a
"Requires:" line with no version information.

Also skip the version restriction for ranges that include '||'
alternatives. These could be handled in the future using RPM's boolean
dependency syntax.

dashea committed:
Fix the handling of semver ranges that contain '*'

Use the semver library to parse version strings into ranges. For any
version strings that fail to parse or are returned as '*', output a
"Requires:" line with no version information.

Also skip the version restriction for ranges that include '||'
alternatives. These could be handled in the future using RPM's boolean
dependency syntax.

ekohl commented on lib/spec_file_generator.js:
```diff
@@ -43,16 +43,19 @@ function dependenciesToRequires(deps) {
-      dependencies.push('npm(' + dep + ') = ' + version);
+      // we should now have a space-separated list of comparators and version numbers
+      // that can be added to a spec file as-is
+      versionRange.split(' ').forEach((constraint) => {
```
How does this handle `>= 1.0 < 2.0`? Wouldn't it end up as 4 entries?

dashea replied:
Granted, this is dependent on the behavior of semver. At least with the current version, the string that validRange spits out is '>=1.0 <2.0', so there's only one space in there. Tests would help to at least ensure that it keeps working.

What I thought would be nice is to have a library that can parse a semver string into an array of objects describing the constraints, but I figured this would be good for now as a quick fix.

ekohl replied:
I'd be OK with it when there are tests

ekohl reviewed (commented):
Thanks, this part was needing some love. Do you know if it also covers `6.x`? Could you write some tests for this?

dashea commented:
> Do you know if it also covers `6.x`

Yes, `semver.validRange("6.x")` returns ">=6.0.0 <7.0.0"

> Could you write some tests for this?

Sure, I'll try to get something put together later this week

ekohl commented:
Great, this crosses off one TODO from my list.

dashea committed:
Fix the handling of exact versions.

Fix the case of exact version requirements, such as '=1.2.3'. Also
rearrange the handling of versionRange results to make the and/or split
more clear and to make it easier to potentially use boolean requirements
in the future.

dashea committed:
Add tests for dependenciesToRequires

dashea commented:
Added tests based on the semver examples and fixed the bug it uncovered (didn't correctly handle constraints equal to one version).

ekohl commented:
Thanks!
